### ISSUE TITLE
HMR protocol no longer correct after Nuxt update; Nuxt vite config not working either

### ISSUE THREAD

**Issue by plumduffer** [nuxt/nuxt#28082]
### Environment

------------------------------
- Operating System: Linux
- Node Version:     v20.15.0
- Nuxt Version:     3.12.3
- CLI Version:      3.12.0
- Nitro Version:    2.9.7
- Package Manager:  pnpm@8.15.4
- Builder:          -
- User Config:      alias, modules, devtools, nitro, shadcn, tailwindcss, runtimeConfig, compatibilityDate
- Runtime Modules:  @nuxtjs/tailwindcss@6.12.0, shadcn-nuxt@0.10.4, nitro-cloudflare-dev@0.1.5
- Build Modules:    -
------------------------------

### Reproduction

This is for my personal website which is public, this is the commit.

Nuxt project is in the frontend folder.

https://github.com/plumduffer/plumduffer-site/tree/f2b14898ea9ad2178a0828060c69365f59e43a4a/frontend

### Describe the bug

After updating Nuxt (^3.11.2  → ^3.12.3), my HMR has stopped working due to it using the wrong protocol.
```
Mixed Content: The page at 'https://plumduffer.lndo.site/password' was loaded over HTTPS, but attempted to connect to the insecure WebSocket endpoint 'ws://plumduffer.lndo.site/_nuxt/'. This request has been blocked; this endpoint must be available over WSS.
```

I am using Lando (https://docs.lando.dev/getting-started/) to setup my local environment inside WSL.

I also tried adding...
```js
    vite: {
        server: {
            hmr: {
                protocol: "wss",
            },
        },
    },
```

...inside my nuxt config, but this does nothing. What's going on here?

**Comment by roberto910907** [comment]
@plumduffer I fixed the issue by adding this to my `nuxt.config.ts`:
```
hooks: {
  'vite:extend'({ config }) {
    if (config.server && config.server.hmr) {
      config.server.hmr.protocol = 'wss'
    }
  },
},
```
The [PR](https://github.com/nuxt/cli/pull/436) in nuxi (nuxt-cli) should resolve the issue by updating the precedence of the protocol key. However, we'll need to wait for the next library release (we're currently using 3.12.0).

Credit for this solution goes to [this comment](https://github.com/nuxt/nuxt/issues/27558#issuecomment-2182901776).

**Comment by db-lyon** [comment]
I've been struggling on this for 2 days now.

As others have done, I added the hook to my `nuxt.config.ts` and it has moved the needle re: `host` and `protocol` but no matter how I go about this i can't get it to respect my `port`. 

Somewhat at a loss so I've disabled HMR for now.

**Comment by chrisspiegl** [comment]
This appears to be the same issue as https://github.com/nuxt/nuxt/issues/27558, and is reported to be fixed in the next nuxi release. Though, I do not know when that will be.

For the time being, comment https://github.com/nuxt/nuxt/issues/27558#issuecomment-2190072771 appears to be a “workaround” until the next nuxi release.

> When using pnpm as package manager, you'll probably need to add this to your `package.json`, using this notation:
> 
> ```json
> "pnpm": {
>     "overrides": {
>       "nuxt>nuxi": "3.11.1"
>     }
>   }
> ```

**Comment by 719media** [comment]
I am using nuxi 3.13.0 and still running into this error message

**Comment by danielroe** [comment]
@719media What is your Nuxt config?

**Comment by 719media** [comment]
There are multiple issues reporting similar issues. Currently I fix the error with an idea from:
https://github.com/nuxt/nuxt/issues/12003#issuecomment-2271609175

```
export default defineNuxtConfig({
  compatibilityDate: '2024-07-02',
  typescript: {
    typeCheck: true,
  },
  modules: ['@pinia/nuxt'],
  hooks: {
    'vite:extendConfig': (config) => {
      config.server = {
        ...config.server,
        hmr: {
          protocol: undefined,
          clientPort: 443,
        },
      };
    },
  },
});
```

when removing the hook, I get the error 
`client:535 Mixed Content: The page at 'DOMAIN' was loaded over HTTPS, but attempted to connect to the insecure WebSocket endpoint 'ws://DOMAIN/_nuxt/'. This request has been blocked; this endpoint must be available over WSS.`

I'm hitting my endpoint over nginx proxy using https in my browser, so the HMR should be using wms by correlation, but it is not, and so I don't get HMR without that hook.

Package versions are
nuxi 3.13.0
nuxt 3.13.0
vue 3.4.38

**Comment by roberto910907** [comment]
@719media Have you tried something like this?
```
export default defineNuxtConfig({
  vite: {
    server: {
      hmr: {
        protocol: 'wss',
      },
    },
  },
....
```
I had to run `npx nuxi upgrade -f` first, but you seem to already have the latest version of nuxi.

**Comment by 719media** [comment]
@roberto910907 Yes, that's fine, I can do that, I was just under the impression that I shouldn't have to configure anything at all (e.g. the hmr protocol wss should automatically match the client protocol https). If it's intended that explicitly stating the protocol is required when doing reverse proxy, that's fine, it just sounded like it was supposed to be more "automatic."

**Comment by danielroe** [comment]
Nuxt shares the main port of your app with the vite websocket, and if you don't launch `nuxi` with `--https` it doesn't realise that you need a secure protocol, so you'll need to provide the configuration as above.

Glad it's working for you when doing that.

**Comment by 719media** [comment]
Thanks. I think the reason it was unexpected to me was because it works fine for me with vue + vite, and I only experience this problem in nuxt. I'll just use the configuration as defined by https://github.com/nuxt/nuxt/issues/28082#issuecomment-2308636640 as it seems the most straightforward.
